### PR TITLE
Play Music From List

### PR DESCRIPTION
**Pull request by IamHaddad**

### CODE REVIEW

File: app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListAdapter.kt [hunk 1]
```diff
@@ -30,7 +32,25 @@ class MusicListItemViewHolder(private val mBinding: ItemMusicBinding) :
     override fun onBind(obj: Music) {
         mBinding.txtMusicTitle.text = obj.title
         mBinding.txtArtist.text = obj.artist
+
+        sendToPlay(obj)
+    }
+
+    private fun sendToPlay(obj: Music) {
+        mBinding.root.setOnClickListener {
+            val bundle = bundleOf(
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListAdapter.kt]: lotfan in karo dakhele fragment anjma bedid
general tip: lotfan kare ezafi toye viewholder hato anjam nadid 
dir ya zod pashimon mishid

File: app/src/main/java/com/worldsnas/starplayer/view/PlayerFragment.kt [hunk 2]
```diff
@@ -69,4 +65,16 @@ class PlayerFragment : Fragment() {
         releaseExoPlayer()
         viewBinding = null
     }
+
+    private fun setMusicArguments() {
+        viewBinding?.tvMusicTitle?.text = arguments?.getString("title")
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/view/PlayerFragment.kt]: ye data class besazi ke in parametr haro negah dare
baad class ro parcelable befrestid be in fragment
injori kheili shirin tare zendegi
bazam jaye tozih dare

File: app/src/main/java/com/worldsnas/starplayer/view/PlayerFragment.kt [hunk 3]
```diff
@@ -33,23 +31,21 @@ class PlayerFragment : Fragment() {
...
+        daggerSetup()
     }
 
     private fun initializeExoplayer() {
         exoPlayer = SimpleExoPlayer.Builder(context!!).build()
         viewBinding?.exoControlView?.player = exoPlayer
 
-        val filePath = Environment.getExternalStorageDirectory().path + "/Download/Temp/Test.mp3"
-        val mediaSource = buildMediaSource(Uri.fromFile(File(filePath)))
+        val filePath = "content://media/" + arguments?.getString("address")
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/view/PlayerFragment.kt]: che jaleb in bayad ezafe she ta kar kone?
bedone `content` nemishe?
ajibe

**Review by helloseyedjafari — APPROVED** [review]
Good job.
just a bit of clean up

**Review by helloseyedjafari — CHANGES_REQUESTED** [review]
eshtebahi approve kardam

File: app/src/main/java/com/worldsnas/starplayer/view/PlayerFragment.kt (hunk 3 above)
Comment (IamHaddad) [reply app/src/main/java/com/worldsnas/starplayer/view/PlayerFragment.kt]: kheir moteasefane nemishe, behtar nist ye variable tarif konam barash?
ya bezaram to model?

File: app/src/main/java/com/worldsnas/starplayer/view/PlayerFragment.kt (hunk 2 above)
Comment (IamHaddad) [reply app/src/main/java/com/worldsnas/starplayer/view/PlayerFragment.kt]: yani bundle moshkel dare?
ye modle dg besazam?

File: app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListAdapter.kt (hunk 1 above)
Comment (IamHaddad) [reply app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListAdapter.kt]: chtor bayad ba ye function bargardonm on music mored nazar ro?

File: app/src/main/java/com/worldsnas/starplayer/view/PlayerFragment.kt (hunk 2 above)
Comment (FatemehMsp) [reply app/src/main/java/com/worldsnas/starplayer/view/PlayerFragment.kt]: mitonid az "safe args" navigation component estefadeh konid kheili karo rahat mikoneh
inam link khode google

https://developer.android.com/guide/navigation/navigation-pass-data#override_a_destination_argument_in_an_action

File: app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListAdapter.kt (hunk 1 above)
Comment (FatemehMsp) [reply app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListAdapter.kt]: mitonid ye interface to adapter bzarid va to fragment azesh estefadeh konid

File: app/src/main/java/com/worldsnas/starplayer/view/PlayerFragment.kt (hunk 2 above)
Comment (helloseyedjafari) [reply app/src/main/java/com/worldsnas/starplayer/view/PlayerFragment.kt]: na bundle moshkeli nadare
farghi ham nemikone az safeargs estefade kardid ya na
ye kelas bepichid dore in parametr haye ersali va on kelas ro pass bedid bejaye khode parameter ha

File: app/src/main/java/com/worldsnas/starplayer/view/PlayerFragment.kt (hunk 3 above)
Comment (helloseyedjafari) [reply app/src/main/java/com/worldsnas/starplayer/view/PlayerFragment.kt]: na bashe felan
vali const bashe behtare

File: app/src/main/java/com/worldsnas/starplayer/ConstValues.kt [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package com.worldsnas.starplayer
+
+object ConstValues {
+
+    const val musicInfoBundleKey = "musicInfo"
+    const val preAddress = "content://media/"
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/ConstValues.kt]: const ha mamolan intori neveshte mishan:
`BUNDLE_KEY_MUSIC_INFO`
albate enum ham injori minevisan

File: app/src/main/AndroidManifest.xml [hunk 5]
```diff
@@ -3,6 +3,7 @@
     package="com.worldsnas.starplayer">
 
     <uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE"/>
+    <uses-permission android:name="android.permission.ACCESS_MEDIA_LOCATION"/>
```
Comment (helloseyedjafari) [inline app/src/main/AndroidManifest.xml:6]: android o majarahash

File: app/src/main/java/com/worldsnas/starplayer/model/Music.kt [hunk 6]
```diff
@@ -1,10 +1,14 @@
 package com.worldsnas.starplayer.model
 
+import android.os.Parcelable
+import kotlinx.android.parcel.Parcelize
+
+@Parcelize
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/model/Music.kt]: afarin

Comment (helloseyedjafari) [reply app/src/main/java/com/worldsnas/starplayer/model/Music.kt]: vali lotfan ye modele jadid besazi
esmesh ham mitonid `LocalMusic` ya chizi shabihesh bezarid
in aslan kare khobi nist ye object az payin tarin laye ta balatarin laye ro control kone

File: app/src/main/java/com/worldsnas/starplayer/view/PlayerFragment.kt [hunk 7]
```diff
@@ -12,19 +11,23 @@ import com.google.android.exoplayer2.source.MediaSource
...
  */
 class PlayerFragment : Fragment() {
 
-    lateinit var playerComponent: PlayerComponent
-    var exoPlayer: SimpleExoPlayer? = null
-    var viewBinding: FragmentPlayerBinding? = null
+    private lateinit var playerComponent: PlayerComponent
+    private var exoPlayer: SimpleExoPlayer? = null
+    private var viewBinding: FragmentPlayerBinding? = null
+    private val music by lazy {
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/view/PlayerFragment.kt]: benazaram ye `lateinit var` sade kheili behtare
lazy overhead e ziadi dare

Comment (helloseyedjafari) [reply app/src/main/java/com/worldsnas/starplayer/view/PlayerFragment.kt]: typesh ham nullable nabashe kheili behtare
mantegh:
in class bedone in object nemitone kar kone 
pas age kasi in object ro yadesh raft befreste ma `throw` konim kheili rahat mishe control va peyda kard moshkel ro

File: app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListAdapter.kt [hunk 8]
```diff
@@ -24,12 +24,20 @@ class MusicsListAdapter :
     }
 }
 
-class MusicListItemViewHolder(private val mBinding: ItemMusicBinding) :
+class MusicListItemViewHolder(private val mBinding: ItemMusicBinding,val onItemClick: ((Music) -> Unit)?) :
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListAdapter.kt]: `onItemClick` nullable che mani mide?
chera bayad null bashe?
benazaram chon niaze pas nemitone null bashe

File: app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListAdapter.kt [hunk 9]
```diff
@@ -11,11 +10,12 @@ import com.worldsnas.starplayer.model.Music
 
 class MusicsListAdapter :
     ListAdapter<Music, MusicListItemViewHolder>(DiffCallback) {
+    var onItemClick: ((Music) -> Unit)? = null
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListAdapter.kt]: inro inja mishe to constructor adapter daryaft kard
va mitone nullable ham nabashe

File: app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListFragment.kt [hunk 10]
```diff
@@ -129,5 +132,21 @@ class MusicsListFragment : Fragment() {
 
     }
 
+    private fun provideMusic() {
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListFragment.kt]: esme behtari ham mishe goazsht

File: app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListFragment.kt [hunk 11]
```diff
@@ -129,5 +132,21 @@ class MusicsListFragment : Fragment() {
...
+
+        musicListAdapter.onItemClick = {
+            val bundle = Bundle()
+
+            bundle.putParcelable(ConstValues.musicInfoBundleKey, it)
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListFragment.kt]: mahze etela ye function has bename:
`bundleOf()` hast

File: app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListFragment.kt [hunk 12]
```diff
@@ -129,5 +132,21 @@ class MusicsListFragment : Fragment() {
...
+
+            _binding!!.root.findNavController()
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListFragment.kt]: baraye find kardan vaghean niaz be view hast?
ro khode fragment fek konam beshe

**Review by helloseyedjafari — CHANGES_REQUESTED** [review]
well done

File: app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListAdapter.kt (hunk 9 above)
Comment (IamHaddad) [reply app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListAdapter.kt]: kodom behtare , goftm dast be constructor nazanam shayad behtar bashe

File: app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListFragment.kt (hunk 11 above)
Comment (IamHaddad) [reply app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListFragment.kt]: nemidonestam mishe ba oon tabe ham parcell ferestad, qablesh ba hamin boode

File: app/src/main/AndroidManifest.xml (hunk 5 above)
Comment (IamHaddad) [reply app/src/main/AndroidManifest.xml:6]: android va mokafat, bedoune in permission roye emulator javab midad

File: app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListAdapter.kt (hunk 8 above)
Comment (IamHaddad) [reply app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListAdapter.kt]: doroste

File: app/src/main/java/com/worldsnas/starplayer/view/PlayerFragment.kt (hunk 7 above)
Comment (IamHaddad) [reply app/src/main/java/com/worldsnas/starplayer/view/PlayerFragment.kt]: try catch khobe?

File: app/src/main/java/com/worldsnas/starplayer/view/PlayerFragment.kt [hunk 13]
```diff
@@ -69,4 +69,20 @@ class PlayerFragment : Fragment() {
         releaseExoPlayer()
         viewBinding = null
     }
+
+    private fun setMusicArguments() {
+        music = arguments?.getParcelable(ConstValues.BUNDLE_KEY_MUSIC_INFO)
+            ?: throw Throwable("Did not Receive Anything ")
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/view/PlayerFragment.kt:75]: afarin
in kheili ali shod

File: app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListAdapter.kt [hunk 14]
```diff
@@ -1,21 +1,21 @@
...
 import android.view.ViewGroup
 import androidx.recyclerview.widget.DiffUtil
 import androidx.recyclerview.widget.ListAdapter
 import com.worldsnas.starplayer.core.BaseViewHolder
 import com.worldsnas.starplayer.databinding.ItemMusicBinding
+import com.worldsnas.starplayer.model.LocalMusic
 import com.worldsnas.starplayer.model.Music
 
-class MusicsListAdapter :
+class MusicsListAdapter(var onItemClick: (LocalMusic) -> Unit) :
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListAdapter.kt]: `private` bashe behtare nist?
vali dar kol cheghad rahat tar shod o shirin?

File: app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListFragment.kt [hunk 15]
```diff
@@ -126,7 +125,13 @@ class MusicsListFragment : Fragment() {
             }
             else -> return
         }
+    }
+
+    private fun musicListener(localMusic: LocalMusic) {
+        val bundle = bundleOf(ConstValues.BUNDLE_KEY_MUSIC_INFO to localMusic)
 
+        findNavController()
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/view/musics_list/MusicsListFragment.kt:133]: hadsam dorost bod pas!

File: app/src/main/java/com/worldsnas/starplayer/model/LocalMusic.kt [hunk 16]
```diff
@@ -0,0 +1,14 @@
+package com.worldsnas.starplayer.model
+
+import android.os.Parcelable
+import kotlinx.android.parcel.Parcelize
+
+@Parcelize
+data class LocalMusic(
```
Comment (helloseyedjafari) [inline app/src/main/java/com/worldsnas/starplayer/model/LocalMusic.kt:7]: khobe dige
baadan dastemon kheili baz tare

**Review by helloseyedjafari — APPROVED** [review]
kheili ali shode kheili daste shoma dard nakone

**Pull request merged**